src/python/pants/goal/run_tracker.py

```diff
@@ -414,21 +414,77 @@ def shutdown_worker_pool(self):
     """
     SubprocPool.shutdown(self._aborted)
 
-  def report_target_info(self, scope, target, key, val):
+  @staticmethod
+  def create_dict_with_nested_keys_and_val(all_keys, value, index):
+    """Recursively constructs a nested dictionary with the keys pointing to the value.
+
+    :param list of string all_keys: A list of keys to be nested as a dictionary.
+    :param dict or string value: The value of the information being stored.
+    :param int index: The index of the last element of the list of keys.
+    :return: dict of nested keys leading to the value.
+    """
+    if index > 0:
+      new_val = {all_keys[index]: value}
+      return RunTracker.create_dict_with_nested_keys_and_val(all_keys, new_val, index - 1)
+    elif index == 0:
+      return {all_keys[index]: value}
+
+  @staticmethod
+  def merge_list_of_keys_into_dict(data, keys, value, index=0):
+    """Recursively merge list of keys that points to the given value into data.
+
+    Will override a primitive value with another primitive value, but will not
+    override a primitive with a dictionary.
+
+    :param dict data: Dictionary to be updated.
+    :param list of string keys: The keys that point to where the value should be stored.
+           Will recursively find the correct place to store in the nested dicts.
+    :param primitive value: The value of the information being stored.
+    :param int index: The index into the list of keys (starting from the beginning).
+    """
+    if len(keys) == 0 or index < 0 or index >= len(keys):
+      raise ValueError('Keys must contain at least one key and index must be'
+                       'an integer greater than 0 and less than the number of keys.')
+    if len(keys) < 2 or not data:
+      new_data_to_add = RunTracker.create_dict_with_nested_keys_and_val(keys, value, len(keys) - 1)
```

Comment:
Again, if this is a @classmethod then these calls to `RunTracker.create_dict_with_nested_keys_and_val` can be replaced with slightly neater calls to `cls.create_dict_with_nested_keys_and_val`